security/rbac/engine/engine.go

```diff
@@ -0,0 +1,187 @@
+			return
+		}
+	}
+	policyName = ""
+	return
+}
+
+// CelEvaluationEngine is the struct for CEL engine.
+type CelEvaluationEngine struct {
+	engines []rbacEngine
+}
+
+// NewCelEvaluationEngine builds a cel evaluation engine from a list of Envoy RBACs.
+func NewCelEvaluationEngine(rbacs []pb.RBAC) (CelEvaluationEngine, error) {
+	if len(rbacs) < 1 || len(rbacs) > 2 {
+		return CelEvaluationEngine{}, status.Errorf(codes.InvalidArgument, "must provide 1 or 2 RBACs")
+	}
+	if len(rbacs) == 2 && (rbacs[0].Action != pb.RBAC_DENY || rbacs[1].Action != pb.RBAC_ALLOW) {
+		return CelEvaluationEngine{}, status.Errorf(codes.InvalidArgument, "when providing 2 RBACs, must have 1 DENY and 1 ALLOW in that order")
+	}
+	var engines []rbacEngine
+	for _, rbac := range rbacs {
+		engines = append(engines, newRbacEngine(rbac))
+	}
+	return CelEvaluationEngine{engines}, nil
+}
+
+// Evaluate is the core function that evaluates whether an RPC is authorized.
+//
+// ALLOW policy. If one of the RBAC conditions is evaluated as true, then the
+// CEL engine evaluation returns allow. If all of the RBAC conditions are
+// evaluated as false, then it returns deny. Otherwise, some conditions are
+// false and some are unknown, it returns undecided.
+//
+// DENY policy. If one of the RBAC conditions is evaluated as true, then the
+// CEL engine evaluation returns deny. If all of the RBAC conditions are
+// evaluated as false, then it returns allow. Otherwise, some conditions are
+// false and some are unknown, it returns undecided.
+//
+// DENY policy + ALLOW policy. If one of the conditions in the DENY policy is
+// evaluated as true, then the CEL engine evaluation returns deny. Otherwise,
+// if one of the conditions in the ALLOW policy is evaluated as true, it
+// returns allow. If all the conditions are evaluated as false, it returns
+// deny. If some conditions are unknown whereas the other conditions are
+// false, it returns undecided.
+func (celEngine CelEvaluationEngine) Evaluate(args AuthorizationArgs) AuthorizationDecision {
+	allUnknownPolicies := ""
+	for _, engine := range celEngine.engines {
+		policyName, unknownPolicies := engine.matches(args)
+		// If any engine matched, return that engine's action.
+		if policyName != "" {
+			var decision Decision
+			if engine.action == pb.RBAC_ALLOW {
+				decision = DecisionAllow
+			} else {
+				decision = DecisionDeny
+			}
+			return AuthorizationDecision{decision, policyName}
+		}
+		allUnknownPolicies += unknownPolicies
+	}
+	// If any engine evaluated to unknown, return unknown.
+	if allUnknownPolicies != "" {
+		return AuthorizationDecision{DecisionUnknown, strings.TrimRight(allUnknownPolicies, ", ")}
+	}
+	// If all engines explicitly did not match.
+	var decision Decision
+	if len(celEngine.engines) == 2 {
+		decision = DecisionDeny
+	} else if celEngine.engines[0].action == pb.RBAC_ALLOW {
+		decision = DecisionDeny
+	} else {
+		decision = DecisionAllow
+	}
+	return AuthorizationDecision{decision, ""}
```

Comment:
One more thing for the above if-else statement is, from a security's perspective, if we have to choose between falsely denying or falsely allowing something, I think the former is better because the latter might introduce some potential attacks. So I would suggest moving the allow decision to the top, like:
```
if len(celEngine.engines) == 1 && celEngine.engines[0].action == pb.RBAC_DENY {
    return AuthorizationDecision{DecisionAllow, ""}    
} 
return AuthorizationDecision{DecisionDeny, ""}
```

Reply:
Ah okay, makes sense. Thanks!